File: server/coolify.js
const { MOCK_COOLIFY_PROJECTS, MOCK_COOLIFY_DEPLOYMENTS, MOCK_CONTAINERS } = require("./mock-data")

const COOLIFY_URL   = process.env.COOLIFY_API_URL
const COOLIFY_TOKEN = process.env.COOLIFY_API_TOKEN

let useRestApi = !!(COOLIFY_URL && COOLIFY_TOKEN)

/**
 * Returns Coolify projects (apps + databases + services).
 * Priority: REST API → Docker label scan → mock data.
 * @param {import('dockerode')|null} docker
 * @returns {Promise<object[]>}
 */
async function getCoolifyProjects(docker = null) {
  if (useRestApi) {
    try {
      const { default: fetch } = await import("node-fetch").catch(() => ({ default: global.fetch }))
      const res = await fetch(`${COOLIFY_URL}/api/v1/projects`, {
        headers: { Authorization: `Bearer ${COOLIFY_TOKEN}`, Accept: "application/json" },
        signal: AbortSignal.timeout(5000),
      })
      if (res.ok) return res.json()
      useRestApi = false
    } catch (err) {
      console.warn("[coolify] REST API failed, falling back:", err.message)
      useRestApi = false
    }
  }

  if (docker) {
    try {
      return await detectFromLabels(docker)
    } catch (err) {
      console.warn("[coolify] Label detection failed:", err.message)
    }
  }

  return MOCK_COOLIFY_PROJECTS
}

/**
 * Returns recent Coolify deployment history.
 * @returns {Promise<object[]>}
 */
async function getCoolifyDeployments() {
  if (!useRestApi) return MOCK_COOLIFY_DEPLOYMENTS
  try {
    const { default: fetch } = await import("node-fetch").catch(() => ({ default: global.fetch }))
    const res = await fetch(`${COOLIFY_URL}/api/v1/deployments`, {
      headers: { Authorization: `Bearer ${COOLIFY_TOKEN}` },
      signal: AbortSignal.timeout(5000),
    })
    if (res.ok) return res.json()
  } catch {}
  return MOCK_COOLIFY_DEPLOYMENTS
}

/**
 * Detect Coolify resources by reading Docker container labels.
 * @param {import('dockerode')} docker
 * @returns {Promise<object[]>}
 */
async function detectFromLabels(docker) {
  const containers = await docker.listContainers({ all: true })
  const projectMap = {}

  for (const c of containers) {
    const labels = c.Labels || {}
    if (!labels["coolify.managed"] && !isCoolifyName(c.Names?.[0] || "")) continue

    const projectName = labels["coolify.projectName"] || labels["coolify.project"] || "coolify"
    const type = labels["coolify.type"] || guessType(c.Image || "")
    const appName = labels["coolify.name"] || (c.Names?.[0] || "").replace(/^\//, "") || c.Id.slice(0, 12)
    const status = c.State === "running" ? "running" : "stopped"

    if (!projectMap[projectName]) {
      projectMap[projectName] = { id: projectName, name: capitalize(projectName), apps: [], databases: [], services: [] }
    }

    const base = { id: c.Id.slice(0, 12), name: appName, status }

    if (type === "database") {
      projectMap[projectName].databases.push({
        ...base, engine: guessEngine(c.Image || ""), size: "—", conns: 0,
      })
    } else if (type === "service") {
      const ports = c.Ports.filter(p => p.PublicPort).map(p => String(p.PublicPort))
      projectMap[projectName].services.push({ ...base, type: guessServiceType(c.Image || ""), ports })
    } else {
      const domains = parseDomains(labels)
      projectMap[projectName].apps.push({
        ...base,
        domains,
        lastDeployed: "—",
        branch: labels["coolify.branch"] || "main",
        containerName: appName,
      })
    }
  }

  return Object.values(projectMap).length > 0 ? Object.values(projectMap) : MOCK_COOLIFY_PROJECTS
}

/**
 * Annotate a container list with Coolify metadata derived from name patterns.
 * @param {object[]} containers
 * @returns {object[]}
 */
function enrichContainersWithCoolify(containers) {
  return containers.map(c => {
    if (c.coolify) return c
    if (!isCoolifyName(c.name)) return c
    return {
      ...c,
      coolify: {
        applicationId: c.id,
        projectName: "coolify",
        appName: c.name,
        type: c.name.match(/db|redis|postgres|mysql|mongo/) ? "database"
          : c.name.match(/proxy|traefik|nginx|caddy|sentinel|realtime|minio/) ? "service"
          : "application",
      },
    }
  })
}

function isCoolifyName(name) {
  const n = name.replace(/^\//, "")
  return n.startsWith("coolify") || /^[a-z0-9]{20,}$/.test(n)
}

function guessType(image) {
  if (/postgres|mysql|redis|mongo|mariadb|clickhouse/.test(image)) return "database"
  if (/traefik|nginx|caddy|minio|sentinel|realtime/.test(image)) return "service"
  return "application"
}

function guessEngine(image) {
  if (image.includes("postgres")) return "postgres"
  if (image.includes("redis"))    return "redis"
  if (image.includes("mysql") || image.includes("mariadb")) return "mysql"
  if (image.includes("mongo"))    return "mongodb"
  if (image.includes("clickhouse")) return "clickhouse"
  return "unknown"
}

function guessServiceType(image) {
  if (image.includes("traefik"))  return "Traefik"
  if (image.includes("nginx"))    return "Nginx"
  if (image.includes("minio"))    return "MinIO"
  if (image.includes("sentinel")) return "Sentinel"
  if (image.includes("realtime")) return "Soketi"
  return image.split("/").pop()?.split(":")[0] || "unknown"
}

function parseDomains(labels) {
  const raw = labels["caddy"] || labels["traefik.http.routers.app.rule"] || ""
  const matches = raw.match(/`([^`]+)`/g) || []
  return matches.map(m => m.replace(/`/g, ""))
}

function capitalize(s) {
  return s.charAt(0).toUpperCase() + s.slice(1)
}

module.exports = {
  getCoolifyProjects,
  getCoolifyDeployments,
  enrichContainersWithCoolify,
  isApiEnabled: () => useRestApi,
}

File: go-api/handlers/coolify.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"

	"pulsenode/api/docker"
)

type CoolifyHandler struct{ dc *docker.Client }

func NewCoolifyHandler(dc *docker.Client) *CoolifyHandler { return &CoolifyHandler{dc: dc} }

func (h *CoolifyHandler) Projects(w http.ResponseWriter, r *http.Request) {
	projects, err := fetchCoolifyProjects(h.dc)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 200, projects)
}

func (h *CoolifyHandler) Deployments(w http.ResponseWriter, r *http.Request) {
	deps, err := fetchCoolifyDeployments()
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 200, deps)
}

func coolifyBase() string  { return os.Getenv("COOLIFY_API_URL") }
func coolifyToken() string { return os.Getenv("COOLIFY_API_TOKEN") }

func fetchCoolifyProjects(dc *docker.Client) ([]map[string]interface{}, error) {
	base := coolifyBase()
	token := coolifyToken()
	if base == "" || token == "" {
		return mockCoolifyProjects(dc), nil
	}
	client := &http.Client{Timeout: 8 * time.Second}
	req, _ := http.NewRequest(http.MethodGet, base+"/api/v1/projects", nil)
	req.Header.Set("Authorization", "Bearer "+token)
	resp, err := client.Do(req)
	if err != nil {
		return mockCoolifyProjects(dc), nil
	}
	defer resp.Body.Close()
	var projects []map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&projects); err != nil {
		return mockCoolifyProjects(dc), nil
	}
	// Enrich with container labels from Docker
	if !dc.IsMock() {
		if ctrs, err := dc.ListContainers(false); err == nil {
			labelIndex := make(map[string]map[string]string)
			for _, c := range ctrs {
				if appID := c.Labels["coolify.applicationId"]; appID != "" {
					labelIndex[appID] = c.Labels
				}
			}
			// Annotate project apps with container state
			for i, p := range projects {
				if apps, ok := p["applications"].([]interface{}); ok {
					for j, a := range apps {
						if app, ok := a.(map[string]interface{}); ok {
							if id, ok := app["id"].(string); ok {
								if labels, ok := labelIndex[id]; ok {
									app["containerName"] = labels["com.docker.compose.service"]
									apps[j] = app
								}
							}
						}
					}
					projects[i]["applications"] = apps
				}
			}
		}
	}
	return projects, nil
}

func fetchCoolifyDeployments() ([]map[string]interface{}, error) {
	base := coolifyBase()
	token := coolifyToken()
	if base == "" || token == "" {
		return mockCoolifyDeployments(), nil
	}
	client := &http.Client{Timeout: 8 * time.Second}
	req, _ := http.NewRequest(http.MethodGet, base+"/api/v1/deployments", nil)
	req.Header.Set("Authorization", "Bearer "+token)
	resp, err := client.Do(req)
	if err != nil {
		return mockCoolifyDeployments(), nil
	}
	defer resp.Body.Close()
	var deps []map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&deps); err != nil {
		return mockCoolifyDeployments(), nil
	}
	return deps, nil
}

func mockCoolifyProjects(dc *docker.Client) []map[string]interface{} {
	apps := []map[string]interface{}{}
	if !dc.IsMock() {
		if ctrs, err := dc.ListContainers(false); err == nil {
			for _, c := range ctrs {
				if appID := c.Labels["coolify.applicationId"]; appID != "" {
					name := strings.TrimPrefix(c.Names[0], "/")
					apps = append(apps, map[string]interface{}{
						"id":            appID,
						"name":          name,
						"status":        map[string]string{"running": "running", "exited": "stopped"}[c.State],
						"containerName": name,
					})
				}
			}
		}
	}
	if len(apps) == 0 {
		apps = []map[string]interface{}{
			{"id": "mock-app-1", "name": "web-frontend", "status": "running", "domains": []string{"example.com"}, "lastDeployed": "2025-05-01", "branch": "main"},
			{"id": "mock-app-2", "name": "api-server", "status": "running", "domains": []string{"api.example.com"}, "lastDeployed": "2025-05-02", "branch": "main"},
		}
	}
	return []map[string]interface{}{
		{
			"id":           "mock-project-1",
			"name":         "Production",
			"applications": apps,
			"databases":    []interface{}{},
			"services":     []interface{}{},
		},
	}
}

func mockCoolifyDeployments() []map[string]interface{} {
	return []map[string]interface{}{
		{"id": "dep-1", "appName": "web-frontend", "branch": "main", "status": "success", "duration": "1m 23s", "triggeredBy": "git push", "timestamp": fmt.Sprintf("%v", time.Now().Add(-2*time.Hour))},
		{"id": "dep-2", "appName": "api-server", "branch": "main", "status": "success", "duration": "0m 45s", "triggeredBy": "manual", "timestamp": fmt.Sprintf("%v", time.Now().Add(-5*time.Hour))},
	}
}
